### PR TITLE
Feature/fp16 precision fix

### PR DESCRIPTION
**Pull request by knwilliams319**
This PR fixes precision issues when training with the `--fp16` option. The fairseq codebase attempts to be general by using a sum reduction in call to `F.cross_entropy()` inside the `AdaptiveLoss` class. For example, this formulation allows for gradient accumulation (via `--update-freq`) to be applied over multiple batches to simulate a larger batch. However, we have GPUs large enough to push a lot of tokens per batch, so we won't use this option. Unfortunately, when the batch size is large enough, the loss at the beginning of training can be too large to fit in a `torch.float16`, making it impossible to converge. 

This PR pushes changes that quickly and dirtily circumvent this problem. Inside the `FairseqCriterion` class, a new `@staticmethod different_forward_for_train_test()` returns `True` for `AdaptiveLoss` only, allowing a `mean` reduction to be used when `forward()` is called during a training step. This returns the correct loss over a single batch and returns a `sample_size` of 1 to ensure that the gradient is not additionally scaled by the original `sample_size`, as the mean reduction automatically scales the loss by `sample_size`. During a validation/test step, the loss calculation uses the original `forward()`, which uses a sum reduction over the batches in the validation/test set. 

In reality, to make this a feature, I believe this option might still be applicable over multiple batches. For example, during a validation step with `AdaptiveLoss`, I think the true loss could be recovered from the individual mean-losses using the loss value and the sample_size reported in a call to `AdaptiveLoss.reduce_metrics()` For example, over two batches with mean-reduced losses $L_1=\frac{\ell_1}{b_1}, L_2\frac{\ell_2}{b_2}$, we could get the full mean loss with $\frac{b_1 L_1 + b_2 L_2}{b_1+b_2}$. With a careful use of scaling factors and loss storage, we might be able to compute this accurately in fp16 without incurring an `inf` from computing $b_1L_1 + b_2L_2$. 

Furthermore, this might be applicable to multiple loss functions. If this is the case, a new option `--use-mean-reduction` could be created, with behavior that alters a criterion that automatically assumes to use a sum reduction to have the behavior described above. If this isn't clear, the `AdaptiveLoss` class has a line that says `F.cross_entropy(..., reduction = 'sum' if reduce else 'none')`.

### CODE REVIEW

**Review by knwilliams319 — COMMENTED** [review]
After reviewing the changed files, I have re-confirmed all changes.
